DITA OT 3.6 - NullPointerException in custom ANT target using xslt pipeline

We have a custom ANT build file with this target:


  <target name="whr-collect-indexterms.3x" description="Build list of index terms" if="dot3x">
    <property name="args.extract.indexterms.xsl" value="${webhelp.responsive.dir}/xsl/indexterms/extractIndexterms.xsl"/>
    <makeurl property="dita.temp.dir.url" file="${dita.temp.dir}"/>
  	<pipeline>
    <xslt basedir="${dita.temp.dir}" destdir="${dita.temp.dir}" style="${args.extract.indexterms.xsl}" reloadstylesheet="${webhelp.reload.stylesheet}">
      <ditafileset format="dita" processingRole="normal"/>
      <param name="TEMPFOLDER" expression="${dita.temp.dir.url}"/>
      <param name="OUT_EXT" expression="${out.ext}"/>
      <mapper type="regexp" from="^(.*?)$$" to="\1.indexterms"/>
      
      <xmlcatalog>
        <catalogpath path="${xsl.extensions.xml.catalog.file.path}"/>
        <catalogpath path="${dita.plugin.org.dita.base.dir}/catalog-dita.xml"/>
      </xmlcatalog>	
    </xslt>
  	</pipeline>   
    
  	<pipeline>
    <xslt in="${dita.temp.dir}/${user.input.file}" out="${output.dir}/index.xml" style="${webhelp.responsive.dir}/xsl/indexterms/collectIndexterms.xsl" classpathref="dost.class.path" reloadstylesheet="${webhelp.reload.stylesheet}">
      <param name="TEMPFOLDER" expression="${dita.temp.dir}"/>
      <param name="JOB_FILE" expression="${dita.temp.dir}/.job.xml"/>
      <param name="INDEXTERMS_PROPERTIES_URL" expression="${indexterms.properties.url}"/>
      <param name="WEBHELP_PARAMETERS_URL" expression="${webhelp.responsive.parameters.file.url}"/>
      
      
      <xmlcatalog>
        <catalogpath path="${xsl.extensions.xml.catalog.file.path}"/>
        <catalogpath path="${dita.plugin.org.dita.base.dir}/catalog-dita.xml"/>
      </xmlcatalog>	
    </xslt>
  	</pipeline>
    
  </target>

when the target is run I get this NullPointerException:

     Caused by: java.lang.NullPointerException
	at org.dita.dost.ant.ExtensibleAntInvoker.getJob(ExtensibleAntInvoker.java:328)
	at org.dita.dost.ant.ExtensibleAntInvoker.execute(ExtensibleAntInvoker.java:173)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
	at sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
	... 22 more

Looking at that particular line of code:

            store = new StreamStore(job.tempDir, xmlUtils);

the "job" can only be null there. Probably you need to replace job.tempDir=>tempDir

--- a/src/main/java/org/dita/dost/ant/ExtensibleAntInvoker.java
+++ b/src/main/java/org/dita/dost/ant/ExtensibleAntInvoker.java
@@ -325,7 +325,7 @@ public final class ExtensibleAntInvoker extends Task {
             Store store = project.getReference(ANT_REFERENCE_STORE);
             if (store == null) {
                 project.log("Store not found from Ant project reference", Project.MSG_ERR);
-                store = new StreamStore(job.tempDir, xmlUtils);
+                store = new StreamStore(tempDir, xmlUtils);
             }
             project.log("Job not found from Ant project reference", Project.MSG_VERBOSE);
             try {

--- a/src/test/java/org/dita/dost/ant/ExtensibleAntInvokerTest.java
+++ b/src/test/java/org/dita/dost/ant/ExtensibleAntInvokerTest.java
@@ -1,0 +1,53 @@
+/*
+ * This file is part of the DITA Open Toolkit project.
+ *
+ * Copyright 2020 Jarno Elovirta
+ *
+ * See the accompanying LICENSE file for applicable license.
+ */
+
+package org.dita.dost.ant;
+
+import org.apache.tools.ant.Project;
+import org.dita.dost.store.StreamStore;
+import org.dita.dost.util.Job;
+import org.dita.dost.util.XMLUtils;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class ExtensibleAntInvokerTest {
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
+    private Project project;
+    private File tempDir;
+
+    @Before
+    public void setUp() throws IOException {
+        tempDir = folder.newFile();
+        project = new Project();
+        project.setUserProperty("dita.temp.dir", tempDir.getAbsolutePath());
+    }
+
+    @Test
+    public void getJob_witJobReference() throws IOException {
+        final Job job = new Job(tempDir, new StreamStore(tempDir, new XMLUtils()));
+        project.addReference("job", job);
+        final Job act = ExtensibleAntInvoker.getJob(project);
+        assertNotNull(act);
+        assertEquals(job, act);
+    }
+
+    @Test
+    public void getJob_withoutJobReference() {
+        final Job act = ExtensibleAntInvoker.getJob(project);
+        assertNotNull(act);
+    }
+}

